uvicorn/protocols/websockets/websockets_impl.py

```diff
@@ -230,6 +229,12 @@ async def asgi_receive(self):
             return {"type": "websocket.connect"}
 
         await self.handshake_completed_event.wait()
+        if not hasattr(self, "transfer_data_task"):
+            # If the handshake completed and there's no data-transfer task,
+            # that means the handshake failed.
+            assert self.closed_event.is_set()
+            return {"type": "websocket.disconnect", "code": self.initial_response[0]}
```

Comment:
Rather than relying on an implementation detail of the `websockets` library, what do we think of this…?

```suggestion
        if self.closed_event.is_set():
            # Handshake has failed.
            return {"type": "websocket.disconnect", "code": self.initial_response[0]}
```